juliet/src/header.rs

```diff
@@ -0,0 +1,406 @@
+    DuplicateRequest = 9,
+    /// Sent a response for request not in-flight.
+    #[error("response for fictitious request")]
+    FictitiousRequest = 10,
+    /// The dynamic request limit has been exceeded.
+    #[error("request limit exceeded")]
+    RequestLimitExceeded = 11,
+    /// Response cancellation for a request not in-flight.
+    #[error("cancellation for fictitious request")]
+    FictitiousCancel = 12,
+    /// Peer sent a request cancellation exceeding the cancellation allowance.
+    #[error("cancellation limit exceeded")]
+    CancellationLimitExceeded = 13,
+}
+
+/// Frame kind, from the kind byte.
+#[derive(Copy, Clone, Debug, EnumCount, EnumIter, Eq, FromRepr, PartialEq)]
+#[cfg_attr(test, derive(proptest_derive::Arbitrary))]
+#[repr(u8)]
+
+pub enum Kind {
+    /// A request with no payload.
+    Request = 0,
+    /// A response with no payload.
+    Response = 1,
+    /// A request that includes a payload.
+    RequestPl = 2,
+    /// A response that includes a payload.
+    ResponsePl = 3,
+    /// Cancellation of a request.
+    CancelReq = 4,
+    /// Cancellation of a response.
+    CancelResp = 5,
+}
+
+impl Header {
+    /// The size (in bytes) of a header.
+    pub(crate) const SIZE: usize = 4;
+    /// Bitmask returning the error bit of the kind byte.
+    const KIND_ERR_BIT: u8 = 0b1000_0000;
+    /// Bitmask returning the error kind inside the kind byte.
+    const KIND_ERR_MASK: u8 = 0b0000_1111;
+    /// Bitmask returning the frame kind inside the kind byte.
+    const KIND_MASK: u8 = 0b0000_0111;
+
+    /// Creates a new non-error header.
+    #[inline(always)]
+    pub const fn new(kind: Kind, channel: ChannelId, id: Id) -> Self {
+        let id = id.get().to_le_bytes();
+        Header([kind as u8, channel.get(), id[0], id[1]])
+    }
+
+    /// Creates a new error header.
+    #[inline(always)]
+    pub const fn new_error(kind: ErrorKind, channel: ChannelId, id: Id) -> Self {
+        let id = id.get().to_le_bytes();
+        Header([
+            kind as u8 | Header::KIND_ERR_BIT,
+            channel.get(),
+            id[0],
+            id[1],
+        ])
+    }
+
+    /// Parse a header from raw bytes.
+    ///
+    /// Returns `None` if the given `raw` bytes are not a valid header.
+    #[inline(always)]
+    pub const fn parse(mut raw: [u8; Header::SIZE]) -> Option<Self> {
+        // Zero-out reserved bits.
+        raw[0] &= Self::KIND_ERR_MASK | Self::KIND_MASK | Self::KIND_ERR_BIT;
+
+        let header = Header(raw);
+
+        // Check that the kind byte is within valid range.
+        if header.is_error() {
+            if (header.kind_byte() & Self::KIND_ERR_MASK) >= ErrorKind::COUNT as u8 {
+                return None;
+            }
+        } else {
+            if (header.kind_byte() & Self::KIND_MASK) >= Kind::COUNT as u8 {
+                return None;
+            }
+
+            // Ensure the 4th bit is not set, since the error kind bits are superset of kind bits.
```

Comment:
It's the 3rd bit, isn't it?

Reply:
The four lower bits of `kind_byte()` are used if `header.is_error()`, but only the lower three bits if the header is NOT an error. This is due to the fact that there are a lot more errors than kind variants. Setting the fourth bit (e.g. `0b0000_1000`) is a spec violation.

The check here is supposed to take `kind_byte`, AND-ing it with `KIND_MASK`, which is `0b0000_0111`. Thus only the lowest three bits are preserved. The `!=` check ensures that kind bite is equal to `0b0000_0XXX`.

We know the highest four are 0, because the highest indicates error status (and `is_error() == false`). The next three (7th, 6th and 5th) are reserved and always zero. So effectively we are checking whether the 4th bit is zero.